### ISSUE TITLE
Cluster not responding

### ISSUE THREAD

**Issue by simon0191** [flynn/flynn#3809]
My flynn cluster is not working at all. Any operation results in an error. 

- `git push` -> `remote: ERROR: Build failed: exec: job exited with status 22`, 
- `flynn scale web=1` -> unknown_error: Something went wrong 

I think that it has something to do with the database. In my rails app I was getting this error "ActiveRecord::StatementInvalid: PG::Error: ERROR:  cannot execute UPDATE in a read-only transaction" so I wasn't able to perform any insertion/update in the DB.

I read that the cause could be that the replica never actually got promoted, but I executed `SELECT pg_is_in_recovery();` (as suggested in http://stackoverflow.com/questions/13223870/activerecordstatementinvalid-pgerror-error-cannot-execute-update-in-a-rea) and it returned false.

Is there a way to restart the DB or flynn?

**Comment by simon0191** [comment]
In the flynn controller logs appears the following error message several times:

`msg="ERROR: cannot execute INSERT in a read-only transaction (SQLSTATE 25006)" component=controller req_id=84d05fd6-94ca-4bd4-8026-089301546aab`

**Comment by josephglanville** [comment]
Can you attempt to grab the cluster debug logs with `flynn-host collect-debug-info`.

**Comment by simon0191** [comment]
I can't, when I try the server runs out of memory `fatal error: runtime: out of memory`, and since I can't scale down any of the apps I can't free memory up :S

**Comment by simon0191** [comment]
Is there a manual way to grab the cluster debug logs?

**Comment by josephglanville** [comment]
Yes, if you could send an archive of all of the logs in `/var/log/flynn` on each node to logs@flynn.io we can try work out what happened.

**Comment by simon0191** [comment]
Done

**Comment by josephglanville** [comment]
It might take me some time to glean anything useful from these logs. In the mean time you might want to attempt to run `flynn-host fix -n 3`, post the output here if it's unsuccessful in restoring the cluster to an operational state.

**Comment by josephglanville** [comment]
Also the output of `flynn-host ps -a` is likely to be very useful in diagnosing the situation.

**Comment by simon0191** [comment]
Here's the output of `flynn-host ps -a`
https://gist.github.com/simon0191/1023189c9c3275b3ff8385b4ae15325f

**Comment by simon0191** [comment]
I ran `flynn-host fix -n 3` but the problem persists. Here's the output of the command:
```
INFO[01-08|01:25:10] found expected hosts                     n=3
INFO[01-08|01:25:10] ensuring discoverd is running on all hosts
INFO[01-08|01:25:12] checking flannel
INFO[01-08|01:25:12] flannel looks good
INFO[01-08|01:25:12] waiting for discoverd to be available
INFO[01-08|01:25:12] checking for running controller API
INFO[01-08|01:25:12] found running controller API instances   n=2
INFO[01-08|01:25:12] found controller instance, checking critical formations
INFO[01-08|01:25:13] checking status of sirenia databases
INFO[01-08|01:25:13] checking for database state              db=postgres
INFO[01-08|01:25:13] checking sirenia cluster status          fn=FixSirenia service=postgres
INFO[01-08|01:25:13] found running leader                     fn=FixSirenia service=postgres
INFO[01-08|01:25:13] found running instances                  fn=FixSirenia service=postgres count=3
INFO[01-08|01:25:13] getting sirenia status                   fn=FixSirenia service=postgres
INFO[01-08|01:25:13] getting service metadata                 fn=FixSirenia service=postgres
INFO[01-08|01:25:13] getting primary job info                 fn=FixSirenia service=postgres job.id=ip1000145-c143c4cb-6d51-4428-92e2-e3ee35b42d7a
INFO[01-08|01:25:13] getting sync job info                    fn=FixSirenia service=postgres job.id=ip1007108-c301c158-d242-4782-8f0f-6dc51e0b7a2c
INFO[01-08|01:25:13] terminating unassigned sirenia instances fn=FixSirenia service=postgres
INFO[01-08|01:25:13] checking for database state              db=mariadb
INFO[01-08|01:25:13] skipping recovery of db, no state in discoverd db=mariadb
INFO[01-08|01:25:13] checking for database state              db=mongodb
INFO[01-08|01:25:13] skipping recovery of db, no state in discoverd db=mongodb
INFO[01-08|01:25:13] checking for running controller API
INFO[01-08|01:25:13] found running controller API instances   n=2
INFO[01-08|01:25:13] found controller instance, checking critical formations
INFO[01-08|01:25:13] scheduler looks up, moving on
INFO[01-08|01:25:13] cluster fix complete
```

**Comment by josephglanville** [comment]
Hi @simon0191 unfortunately the logs aren't proving to be very enlightening as it has been looping long enough to overwrite both the main log file and the rotated one. Is there any chance I might be able to get a shell to attempt some further debugging?

Failing that it is likely the cluster will recover if you kill the current `sync` peer with `flynn-host stop ip1007108-c301c158-d242-4782-8f0f-6dc51e0b7a2c` this should boot the state machine into action and have it replace the `sync` node.

**Comment by josephglanville** [comment]
Before doing that if you could attempt to collect logs for that process with `flynn-host log ip1007108-c301c158-d242-4782-8f0f-6dc51e0b7a2c` that would be much appreciated too.

**Comment by simon0191** [comment]
Looks like the cluster is operating normally after running`flynn-host stop ip1007108-c301c158-d242-4782-8f0f-6dc51e0b7a2c`, thanks @josephglanville!

How did you know what node to stop? the leader one? and How did you obtained the `ip1007108-c301c158-d242-4782-8f0f-6dc51e0b7a2c` identifier ?

I also collected the logs of `ip1007108-c301c158-d242-4782-8f0f-6dc51e0b7a2c` and emailed to logs@flynn.io.

**Comment by josephglanville** [comment]
The log you just sent through contained critical info, specifically the host running the sync was unable to connect to the primary because of being out of file descriptors. Sadly the log didn't cover the time period before it got stuck looping.

I determined which node to stop from the `flynn-host fix -n3` output, the `sync` node in this case.
However it could also be determined by querying `discoverd`, i.e: `curl http://127.0.0.1/services/discoverd/meta` and finding the `sync` node.

To get to the root cause we need to understand if it was the postgres appliance leaking fds or another process on the system and correct the problem so it doesn't happen again.

**Comment by simon0191** [comment]
- When I run `http://127.0.0.1/services/discoverd/meta` I get `Not Found`.

- Since we had to migrate the service to heroku the flynn cluster is not in use, so, if it's of any help to find the root cause of the problem, I can give you shell access. Give me your email and I'll send you the credentials.

**Comment by josephglanville** [comment]
Oh my bad, I too hastily submitted that without fixing the curl command.
It should be `curl http://127.0.0.1:1111/services/postgres/meta`.

My email is joseph@flynn.io, if you could add my public key available at https://github.com/josephglanville.keys and send me username + endpoint that would be very much appreciated.

**Comment by simon0191** [comment]
Done

**Comment by josephglanville** [comment]
Root cause may have been the zpool running out of space, which caused discoverd to fail.
Seems to have been caused by accumulated connectivity blips, each time that caused a postgres failover a new data volume was allocated and filled up. Over time that lead to running the zpool out of space.
I took the liberty of deleting the old volumes which caused the cluster to resume normal operation promptly.
In future you probably want to deploy machines with bigger zpool for production usage.
Newer versions of Flynn have volume garbage collection tools too so that will help in future.
The too many open file descriptors seems to have been somewhat of a red herring but I am going to look into it anyway.
